libs/lensfun/mount.cpp

```diff
@@ -6,51 +6,82 @@
 #include "config.h"
 #include "lensfun.h"
 #include "lensfunprv.h"
-#include <stdlib.h>
+#include <cstdlib>
+#include <algorithm>
 
-lfMount::lfMount ()
+lfMount::lfMount () :
+     Name{nullptr},
+     Compat{nullptr}
 {
-    Name = NULL;
-    Compat = NULL;
 }
 
 lfMount::~lfMount ()
 {
     lf_free (Name);
-    for (char* m: MountCompat)
-        free(m);
 }
 
-lfMount::lfMount (const lfMount &other)
+lfMount::lfMount (const lfMount &other) :
+     Name{lf_mlstr_dup (other.Name)},
+     Compat{nullptr},
+     MountCompat{other.MountCompat}
 {
-    Name = lf_mlstr_dup (other.Name);
-    Compat = NULL;
+    RebuildPointers();
+}
 
-    MountCompat.clear();
-    if (auto* otherMounts = other.GetCompats())
-    {
-        for (int i = 0; otherMounts[i]; i++)
-            AddCompat(otherMounts[i]);
-    }
+lfMount::lfMount (lfMount &&other) noexcept :
+     Name{std::move(other.Name)},
+     Compat{std::move(other.Compat)},
+     MountCompat{std::move(other.MountCompat)},
+     MountCompatPtrs{std::move(other.MountCompatPtrs)}
+{
+    other.Name = nullptr;
+    other.Compat = nullptr;
 }
 
 lfMount &lfMount::operator = (const lfMount &other)
 {
     lf_free (Name);
     Name = lf_mlstr_dup (other.Name);
-    Compat = NULL;
 
-    MountCompat.clear();
-    if (auto* otherMounts = other.GetCompats())
-    {
-        for (int i = 0; otherMounts[i]; i++)
-            AddCompat(otherMounts[i]);
-    }
+    MountCompat = other.MountCompat;
+
+    RebuildPointers();
 
     return *this;
 }
 
-bool lfMount::operator == (const lfMount& other)
+lfMount &lfMount::operator = (lfMount &&other) noexcept
+{
+    Name = std::move(other.Name);
+    Compat = std::move(other.Compat);
+    MountCompat = std::move(other.MountCompat);
+    MountCompatPtrs = std::move(other.MountCompatPtrs);
+
+    other.Name = nullptr;
+    other.Compat = nullptr;
+
+    return *this;
+}
+
+void lfMount::RebuildPointers()
+{
+    MountCompatPtrs.clear();
+
+    // +1 for C-compatible null-terminated vector
+    MountCompatPtrs.reserve(MountCompat.size() + 1);
+    std::transform(MountCompat.begin(),
+                   MountCompat.end(),
+                   std::back_inserter(MountCompatPtrs),
+                   [](const std::string& mount){ return mount.c_str(); });
+
+    // making the vector C-compatible
+    MountCompatPtrs.emplace_back(nullptr);
```

Comment:
Here we don't use the trick with potential `UB` when accessing `std::vector` after the `end()` in `_lf_terminate_vec`.
Since we don't care about actual `MountCompatPtrs` size we can simply put `nullptr` at the end.